### PR TITLE
Shared Recipe

### PR DESCRIPTION
**Pull request by joeychin01**
Added the ability for users to share recipes with other users

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/ShareRecipeServlet.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        Entity messagesEntity = results.asSingleEntity();
+        if(messagesEntity == null) {
+          System.out.println(userService.getCurrentUser().getEmail());
+        } else {
+          ArrayList<String> messages = convertToList(messagesEntity, "messages");
+          ArrayList<String> recipeIds = convertToList(messagesEntity, "recipeIds");
+          ArrayList<String> userEmails = convertToList(messagesEntity, "userEmails");
+          System.out.println(messages);
+          System.out.println(recipeIds);
+          System.out.println(userEmails);
```
Comment (ALai2) [inline src/main/java/com/google/sps/servlets/ShareRecipeServlet.java:71]: nit: Remove print statements

File: src/main/java/com/google/sps/servlets/ShareRecipeServlet.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import org.json.JSONArray;
+
+/** Servlet that posts and gets user profiles in Datastore */
+@WebServlet("/shareRecipe")
+public class ShareRecipeServlet extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+     UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+    } else {
```
Comment (liguanghua) [inline src/main/java/com/google/sps/servlets/ShareRecipeServlet.java:50]: Use "if (userService.isUserLoggedIn())" when there is nothing to do for users not logged in.

File: src/main/java/com/google/sps/servlets/ShareRecipeServlet.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      Query query = new Query("Profile")
+        .setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, userService.getCurrentUser().getUserId()));
+      PreparedQuery results = datastore.prepare(query);
+      Entity userEntity = results.asSingleEntity();
+      if(userEntity == null) {
+        //no profile
+      } else {
```
Comment (liguanghua) [inline src/main/java/com/google/sps/servlets/ShareRecipeServlet.java:58]: Ditto.

File: src/main/java/com/google/sps/servlets/ShareRecipeServlet.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.setContentType("application/html");
+      response.getWriter().println("Please log in to Share recipes");
+    } else {
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      Query query = new Query("Profile")
+      .setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, userService.getCurrentUser().getUserId()));
```
Comment (liguanghua) [inline src/main/java/com/google/sps/servlets/ShareRecipeServlet.java:99]: +4 indentation

File: src/main/java/com/google/sps/servlets/ShareRecipeServlet.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        String friendEmail = request.getParameter("friendEmail");
+        query = new Query("Profile")
+          .setFilter(new Query.FilterPredicate("email", Query.FilterOperator.EQUAL, friendEmail));
+        results = datastore.prepare(query);
+        Entity friendEntity = results.asSingleEntity();
+        if(friendEntity == null) {
+          response.setContentType("application/html");
+          response.getWriter().println("Sorry, email not found");      
+        }
+        else {
```
Comment (liguanghua) [inline src/main/java/com/google/sps/servlets/ShareRecipeServlet.java:115]: move to the line above.

File: src/main/webapp/script.js [hunk 6]
```diff
@@ -1003,4 +1013,97 @@ function readUserDishInput() {
...
+function loadSentRecipes() {
+  const overlay = document.getElementById('overlay');
+  overlay.style.display = 'block';
+  getLoginStatus('shareRecipe.html');
+  fetch('/shareRecipe').then(response => response.json()).then((messageJsonList) => {
+    const sharedRecipeListElement = document.getElementById("shared-messages");
+    messageJsonList.forEach(sharedRecipe => {
+      createSharedRecipeElement(sharedRecipeListElement, sharedRecipe);
+    });
+  overlay.style.display = 'none';
```
Comment (liguanghua) [inline src/main/webapp/script.js:1084]: +2 indentation.

**Pull request merged**